# PRESCRIPTIVE ANALYTICS [ MUESLI PROJECT ]


#### **WHY PRESCRIPTIVE ANALYTICS?**

We have taken a data-driven approach to find out:

- What is happening --> Transforming Data into Information 

- Why it is happening. --> Transforming Information to Knowledge

Now we will provide recommendations for the "What should we do?"

Transform the knowledge into an action, that will ultimately benefit the client/stakeholder. 





---
#### **OBJECTIVE OF THIS NOTEBOOK:** 

This notebook is the transitionary point from Exploaratory Data Analysis to Explanatory Data Analysis.

We have performed Descriptive Analytics to explore what is happening & Diagnostic Analytics to explore and understand why it is happening.

Here we will use this Exploratory knowledge, the pearls that we have picked from the shells in the vast sea of data, to explain what the impact of our EDA has in the context of the Muesli Warehouse Company. 

 
We will prescribe the following options:

    1.  Daily Trucks for pickup from Warehouse 

    2.  Express Ship mode only

We will still use the KPIs to show the potential impact that these recommendations will have.


---

## Import Packages

In [23]:
# Package for processing data pandas with alias pd
import pandas as pd

# Package to enable us to perform some mathematical calculations
import numpy as np


## Load the data


In [24]:

# order_cycle_kpi
order_cycle_kpi = pd.read_csv('../data/order_cylce_kpi.csv') 
# lead_time_kpi
lead_time_kpi = pd.read_csv('../data/lead_time_kpi.csv')

# shipping_kpi
shipping_kpi = pd.read_csv('../data/shipping_kpi.csv')

# pickup_time_kpi
pickup_time_kpi = pd.read_csv('../data/pickup_time_kpi.csv')

# order_processed_cleaned this contains information on ship mode
order_process = pd.read_csv("../data/order_process_cleaned.csv")

# intern data 
intern_data = pd.read_csv("../data/intern_study_cleaned.csv")

# campaign data
campaign = pd.read_csv("../data/campaign_cleaned.csv")

---
# Recommendations

## Scenarios Daily Trucks for Pickup

Let's re-draw our business flow chart for the scenario of daily truck pickups

![kpi_order_pickup_daily_trucks.png](images/kpi_order_pickup_daily_trucks.png)

In the last session we calculated the actual reality of this.

#### General observations on order pickup time from last session

* We could confirm our assumption that the ship mode really has an effect on the order pickup time.

* The standard processing has 1 day of order pickup time only for 40% of cases, the other orders need 2 or 3 days. 

* The standard processing is delayed in 60% of cases because half of the orders are ready for pick up on Monday which takes two days to be on truck. If we can improve the shipping process for Monday it will have a noticable impact on the order pickup KPI for standard overall.

### Let's first look into Standard

Let's recap the order pickup times for standard processing from the last session:

![order_ready_toship_weekdays_table_overview_standard.png](images/order_ready_toship_weekdays_table_overview_standard.png)

### Ideal Impact on daily truck pickups (for Standard)

- Monday: from 2.0 to 1.0

- Tuesday: same

- Wednesday: from 2.0 to 1.0

- Thursday: same

- Friday: same

### What would be the impact on the order pickup time?

We can solve this by a simple business case analysis ...

Let's calculate the general order pickup time again based on this table.

In [88]:
# general order pickup time - a simple weighted average calculation

# number of order per day
mon_count = 74
tue_count = 24
wed_count = 16
thu_count = 15
fri_count = 22

weighted_sum = (2 * mon_count 
                + 1 * tue_count 
                + 2 * wed_count 
                + 1 * thu_count 
                + 1 * fri_count)
counts_sum = mon_count + tue_count + wed_count + thu_count + fri_count

avg_order_pickup = round(weighted_sum/counts_sum,2)

In [89]:
print(weighted_sum)
print(counts_sum)
print(avg_order_pickup)

241
151
1.6


### Let's take the formula above and adjust it according to our scenario. 


![updated_order_to_ship_standard.png](images/updated_order_to_ship_standard.png)

In [90]:
# solution
# take the formula above and adjust 
new_weighted_sum = (  1 * mon_count 
                    + 1 * tue_count 
                    + 1 * wed_count 
                    + 1 * thu_count 
                    + 1 * fri_count)
new_counts_sum = mon_count + tue_count + wed_count + thu_count + fri_count

new_avg_order_pickup = round(new_weighted_sum/new_counts_sum,2)

In [91]:
print(new_avg_order_pickup)

1.0


#### How much did we improve in percentage?


In [92]:
# solution
improvement_perc = round((new_avg_order_pickup-avg_order_pickup)/avg_order_pickup,2)*100

In [93]:
print(str(improvement_perc) + " %")


-38.0 %


#### **Final outcome** for the impact on standard processing with daily truck pickups: `1 day (-38%)`

## Let's move to Express Shippings

A recap from the previous notebook's findings:

![order_ready_toship_weekdays_table_overview_express.png](images/order_ready_toship_weekdays_table_overview_express.png)


In [94]:
# general order pickup time - a simple weighted average calculation

# number of order per day
mon_count_expr = 19
tue_count_expr = 5
wed_count_expr = 7
thu_count_expr = 13
fri_count_expr = 9

weighted_sum_expr = (  0 * mon_count_expr 
                     + 1 * tue_count_expr 
                     + 0 * wed_count_expr 
                     + 1 * thu_count_expr 
                     + 0 * fri_count_expr)
counts_sum_expr = mon_count_expr + tue_count_expr + wed_count_expr + thu_count_expr + fri_count_expr

avg_order_pickup_expr = round(weighted_sum_expr/counts_sum_expr,2)

In [95]:
print(weighted_sum_expr)
print(counts_sum_expr)
print(avg_order_pickup_expr)

18
53
0.34


### What would be the impact on the order pickup time?

![updated_order_to_ship_express.png](images/updated_order_to_ship_express.png)

- Monday: same

- Tuesday: from 1 to 0

- Wednesday: same

- Thursday: from 1 to 0

- Friday: same

In [96]:
# solution
# take the formula above and adjust 
new_express_totals = (0 * mon_count_expr 
                    + 0 * tue_count_expr 
                    + 0 * wed_count_expr 
                    + 0 * thu_count_expr 
                    + 0 * fri_count_expr)
new_express_counts = mon_count_expr + tue_count_expr + wed_count_expr + thu_count_expr + fri_count_expr

new_express_avg_order_pickup = round(new_express_totals/new_express_counts,2)

In [97]:
print(new_express_totals)
print(new_express_counts)
print(new_express_avg_order_pickup)

0.0
53
0.0


In [98]:
new_express_avg_order_pickup

0.0

#### How much did we improve in percentage?

In [99]:
 # solution
improvement_perc = round((new_express_avg_order_pickup-avg_order_pickup_expr)/avg_order_pickup_expr,2)*100

In [101]:

print(str(improvement_perc) + " %")

-100.0 %


#### **Final outcome** for the impact on express processing with daily truck pickups: `0.0day (-100%)`

### Differences in pickup times - original vs every day:

| **Orders** | **Old pick up time** | **New every day pick up time** |
| --- | --- | --- |
| All ship modes | 1.27 days | 0.74 days |
| Express | 0.34 days | 0.0 days [same day] |
| Standard Processing | 1.6 days | 1.0 days |


----
## Shipping Mode 
In the following we will explore the recommendation that all orders will be treated as Express only. This is a more difficult recommedation to explore.

There are certain aspects of the process that we cannot explain yet.

- Order Cycle KPI : We saw in the last session that this KPI does not meet the warehouse assumptions for both Standard and Express.

- Pickup Time KPI: has a knock on effect due to the difference in Order Cycle. The Standard processed orders are supposed to also wait for 1 day to be shipped after they are ready for pickup.

![]("ad_alliance_mbc/muesli_project/images/order_cycle_time_by_shipmode.png"

![](./3_data_aggregation/static/kpi.png

These 2 KPIs affect the overall Lead Time for the orders. Therefore making a huge impact, we could spend a week easily to come up with expansive and exhaustive process to show the effect on going all Express shipping only.


We will explore a method that will try to show possible impact of going all Express deliveries.

1. **`Step 1:`** Adjust the Order pickup time for standard processing

    in an ideal world we would have a difference of 1 day for the processing time of standard and express. However, we saw that also the order cycle time is affected by the ship mode, so the difference is actual more than 1 day for some orders. In order to account for the actual reality, we will work with the average difference between the processing times of standard and express. 

2. **`Step 2:`** Get the average values for each KPI by ship_mode.

    `standard.mean() & express.mean()`

3. **`Step 3:`** Take the difference between the averages

    `delta = standard.mean()  - express.mean()`

4. **`Step 4:`** Take the resultant and subtract every order with Standard processing in order to arrive at a scenario with all orders with Express.

    `respective_kpi["new_column] = respective_kpi_rows - delta`

5. **`Step 5:`** Take the average of the new column and compare the KPI values.

    `respective_kpi["new_column].mean()`


## Order Cycle Time Shipping method


We will explore the first KPI in the process to determine on average what difference it would make to only use the Express shipping mode


#### Merging and cleaning 

In [42]:
# merge the order_cycle_kpi dataframe with the order_process in order to get the shipping modes
mode_order_cycle_kpi = order_cycle_kpi.merge(order_process, on="order_id")

In [43]:
# Check the columns for "mode_order_cycle_kpi"
mode_order_cycle_kpi.____

Index(['order_id', 'order_date_x', 'origin_channel', 'region', 'category',
       'sub_category', 'ready_to_ship_date', 'pickup_date',
       'expected_ready_to_ship_date', 'order_year', 'order_to_ready',
       'order_to_ready_int', 'order_weekday', 'ready_weekday',
       'order_weekday_int', 'order_to_ready_int_no_weekend', 'order_date_y',
       'on_truck_scan_date', 'ship_mode'],
      dtype='object')

<details>

<summary>Spoiler</summary>

```python

# Check the columns
mode_order_cycle_kpi.columns
               
```
</details>

In [44]:
# rename the order_date_x to order_date
mode_order_cycle_kpi.rename(columns={"order_date_x": "order_date"}, inplace= True)

# drop order_date_y and other columns that are not important in calculation of the time 
mode_order_cycle_kpi.drop(columns=["order_date_y", "origin_channel", "region","category","sub_category"], inplace=True)


In [45]:
# display the first 5 rows of the dataframe "mode_order_cycle_kpi"
mode_order_cycle_kpi.____

order_id  order_date ready_to_ship_date pickup_date  \
0  CA-2020-119004  2020-11-23         2020-11-30  2020-12-02   
1  CA-2019-145625  2019-09-11         2019-09-17  2019-09-18   
2  CA-2020-135307  2020-11-26         2020-11-27  2020-11-27   
3  CA-2019-102981  2019-09-06         2019-09-09  2019-09-11   
4  US-2020-155425  2020-11-10         2020-11-11  2020-11-12   

  expected_ready_to_ship_date  order_year order_to_ready  order_to_ready_int  \
0                  2020-11-25        2020         7 days                   7   
1                  2019-09-13        2019         6 days                   6   
2                  2020-11-28        2020         1 days                   1   
3                  2019-09-08        2019         3 days                   3   
4                  2020-11-12        2020         1 days                   1   

  order_weekday ready_weekday  order_weekday_int  \
0        Monday        Monday                  1   
1     Wednesday       Tuesday                  3   
2      Thursday        Friday                  4   
3        Friday        Monday                  5   
4       Tuesday     Wednesday                  2   

   order_to_ready_int_no_weekend on_truck_scan_date            ship_mode  
0                              5         2020-12-02  Standard Processing  
1                              4         2019-09-18  Standard Processing  
2                              1         2020-11-27              Express  
3                              1         2019-09-11  Standard Processing  
4                              1         2020-11-12              Express

<details>

<summary>Spoiler</summary>

```python

# display the first 5 rows of the dataframe "mode_order_cycle_kpi"
mode_order_cycle_kpi.head()
               
```
</details>

#### Step 1: Splitting by the Ship Mode & Getting Averages


In [46]:
# "express_order_cycle" is a subset of mode_order_cycle_kpi dataframe for orders that are express
express_order_cycle = mode_order_cycle_kpi[mode_order_cycle_kpi["ship_mode"] == "Express"]

In [48]:
# inspect the shape of the "express_order_cycle" dataframe
express_order_cycle._____

(53, 14)

<details>

<summary>Spoiler</summary>

```python

# inspect the shape of the "express_order_cycle" dataframe
express_order_cycle.shape
               
```
</details>

In [47]:
# "standard_order_cycle" is a subset of mode_order_cycle_kpi dataframe for the number of orders that are standard

standard_order_cycle = mode_order_cycle_kpi[mode_order_cycle_kpi["_________"] == "Standard Processing"]

<details>

<summary>Spoiler</summary>

```python

#standard_order_cycle is a subset of mode_order_cycle_kpi dataframe for the number of orders that are standard

standard_order_cycle = mode_order_cycle_kpi[mode_order_cycle_kpi["ship_mode"] == "Standard Processing"]
               
```
</details>

In [49]:
# average order_cycle_time for only express orders
avg_express_order_cycle= express_order_cycle.order_to_ready_int_no_weekend.mean()

In [50]:
# average order_cycle_time for only standard orders

avg_standard_order_cycle = standard_order_cycle.order_to_ready_int_no_weekend.____

<details>

<summary>Spoiler</summary>

```python

# average order_cycle_time for only standard orders

avg_standard_order_cycle = standard_order_cycle.order_to_ready_int_no_weekend.mean()
               
```
</details>

#### Average Order Cycle time per shipping mode

In [51]:
avg_express_order_cycle

1.3962264150943395

In [52]:
avg_standard_order_cycle

3.251655629139073

In [53]:
# exploring if there are instances where the standard processing orders were ready on the same day excluding weekends
standard_order_cycle[standard_order_cycle["order_to_ready_int_no_weekend"] == 0]

order_id  order_date ready_to_ship_date pickup_date  \
6    CA-2020-122504  2020-11-07         2020-11-09  2020-11-12   
13   CA-2020-152737  2020-11-07         2020-11-09  2020-11-12   
170  CA-2020-117702  2020-11-28         2020-11-30  2020-12-02   
199  CA-2020-152142  2020-11-14         2020-11-16  2020-11-18   

    expected_ready_to_ship_date  order_year order_to_ready  \
6                    2020-11-09        2020         2 days   
13                   2020-11-09        2020         2 days   
170                  2020-11-30        2020         2 days   
199                  2020-11-16        2020         2 days   

     order_to_ready_int order_weekday ready_weekday  order_weekday_int  \
6                     2      Saturday        Monday                  6   
13                    2      Saturday        Monday                  6   
170                   2      Saturday        Monday                  6   
199                   2      Saturday        Monday                  6   

     order_to_ready_int_no_weekend on_truck_scan_date            ship_mode  
6                                0         2020-11-12  Standard Processing  
13                               0         2020-11-12  Standard Processing  
170                              0         2020-12-02  Standard Processing  
199                              0         2020-11-18  Standard Processing

There are 4 instances where the "order_ready_int_no_weekend" is 0 days, it will be something to keep in the back of the minds as in step 3 we would be subtracting the delta which we will calculate next.

Can we really have a negative "order_ready_int_no_weekend"??

#### Step 2 : The delta between each ship_mode average

In [54]:
# calculating the delta between the averages, this will be used to subract the order_cycle_time_no_weeekends
delta = avg_standard_order_cycle - avg_express_order_cycle

#### Step 3 : Create new column taking into account the delta


In [ ]:
# Subtract the value only for non-Express shipping method
mode_order_cycle_kpi["expressed_no_weekend"] = np.where(mode_order_cycle_kpi['ship_mode'] != 'Express', #selecting all non-express orders
                                             mode_order_cycle_kpi["order_to_ready_int_no_weekend"] - delta, # subtracting delta from them 
                                             mode_order_cycle_kpi["order_to_ready_int_no_weekend"]) #for all instances that are express, leave the same values

In [56]:
#Investigating if the numbers make sense
mode_order_cycle_kpi[["expressed_no_weekend","order_to_ready_int_no_weekend","ship_mode"]]

expressed_no_weekend  order_to_ready_int_no_weekend            ship_mode
0                3.144571                              5  Standard Processing
1                2.144571                              4  Standard Processing
2                1.000000                              1              Express
3               -0.855429                              1  Standard Processing
4                1.000000                              1              Express
..                    ...                            ...                  ...
199             -1.855429                              0  Standard Processing
200              3.144571                              5  Standard Processing
201              0.144571                              2  Standard Processing
202              3.000000                              3              Express
203              1.144571                              3  Standard Processing

[204 rows x 3 columns]

Clearly they dont for the order instances that take 0 days  "order_to_ready_int_no_weekend"
`i.e. index 199 and 201` 


**We cant process an order that we have not yet even received**



In [57]:
# we take the maximum value for each row or replace it with 1 if less than 1 
mode_order_cycle_kpi["expressed_no_weekend"] = np.maximum(mode_order_cycle_kpi["expressed_no_weekend"], 1)

In [58]:
mode_order_cycle_kpi[["expressed_no_weekend","order_to_ready_int_no_weekend","ship_mode"]]

expressed_no_weekend  order_to_ready_int_no_weekend            ship_mode
0                3.144571                              5  Standard Processing
1                2.144571                              4  Standard Processing
2                1.000000                              1              Express
3                1.000000                              1  Standard Processing
4                1.000000                              1              Express
..                    ...                            ...                  ...
199              1.000000                              0  Standard Processing
200              3.144571                              5  Standard Processing
201              1.000000                              2  Standard Processing
202              3.000000                              3              Express
203              1.144571                              3  Standard Processing

[204 rows x 3 columns]

Aha looking at the 2 row 199 and 201 , the problem is fixed. 

#### Step 4: The new KPI based on only Express ship_mode

In [59]:
# The new average order_cycle if only express using the column "expressed_no_weekend"
new_order_cycle_kpi = mode_order_cycle_kpi.expressed_no_weekend._____

1.702629896141888

<details>

<summary>Spoiler</summary>

```python

# The new average order_cycle if only express
new_order_cycle_kpi = mode_order_cycle_kpi.expressed_no_weekend.mean()

```
</details>

In [ ]:
#check the value 
new_order_cycle_kpi

|  **KPIs**                 | **Warehouse Assumption**|  **Without Weekend**|**Prescription Express** |
|  -----------              |      -----------        |              ----      |        ----             |
|  Order Cycle Time         |          2 days         |         2.80 days  |        1.72 days       |


## Pickup time with Express shipping


We now explore the second KPI, the pickup time and try to evaluate the impact of switching to only express orders

In [60]:
#Using the pickup_time_kpi dataframe., inspect the head
pickup_time_kpi.____

order_id ready_to_ship_date pickup_date  order_date  \
0  CA-2019-116540         2019-09-02  2019-09-03  2019-09-02   
1  CA-2019-129847         2019-09-04  2019-09-04  2019-09-02   
2  CA-2019-129630         2019-09-04  2019-09-04  2019-09-04   
3  CA-2019-106278         2019-09-05  2019-09-06  2019-09-02   
4  CA-2019-158099         2019-09-05  2019-09-06  2019-09-03   

  on_truck_scan_date            ship_mode pickup_time_delta  \
0         2019-09-03              Express            1 days   
1         2019-09-04              Express            0 days   
2         2019-09-04              Express            0 days   
3         2019-09-06  Standard Processing            1 days   
4         2019-09-06              Express            1 days   

   pickup_time_delta_int order_weekday on_truck_weekday ready_to_ship_weekday  \
0                      1        Monday          Tuesday                Monday   
1                      0        Monday        Wednesday             Wednesday   
2                      0     Wednesday        Wednesday             Wednesday   
3                      1        Monday           Friday              Thursday   
4                      1       Tuesday           Friday              Thursday   

   order_weekday_int  ready_to_ship_int  on_truck_int  \
0                  1                  1             2   
1                  1                  3             3   
2                  3                  3             3   
3                  1                  4             5   
4                  2                  4             5   

   pickup_time_delta_no_weekend  
0                             1  
1                             0  
2                             0  
3                             1  
4                             1

<details>

<summary>Spoiler</summary>

```python

#Using the pickup_time_kpi dataframe., inspect the head
pickup_time_kpi.head()

```
</details>

In [61]:
#check the shape
pickup_time_kpi.shape

(204, 15)

#### Step 1: Splitting by the Ship Mode & Getting Averages

In [62]:
# create a subset containing only express orders called express_pickup
express_pickup = pickup_time_kpi[pickup_time_kpi["ship_mode"] == "Express"]

In [ ]:
# create a subset containing only standard orders called standard_pickup from the "pickup_time_kpi" dataframe
standard_pickup = pickup_time_kpi[_______["ship_mode"] != "Express"]


<details>

<summary>Spoiler</summary>

```python

# create a subset containing only standard orders called standard_pickup
standard_pickup = pickup_time_kpi[pickup_time_kpi["ship_mode"] != "Express"]

```
</details>

#### The Average for each ship mode

In [67]:
# calculate the mean pickup time excluding the weekends
avg_express_pickup = express_pickup["pickup_time_delta_no_weekend"].mean()

0.39622641509433965

In [68]:
# calculate the mean pickup time excluding the weekends
avg_standard_pickup = standard_pickup["______"].______()

1.695364238410596

<details>

<summary>Spoiler</summary>

```python

# calculate the mean pickup time excluding the weekends
avg_standard_pickup = standard_pickup["pickup_time_delta_no_weekend"].mean()

```
</details>

#### Step 2 : The delta between each ship_mode average


In [69]:
# calculate the delta for pickup_time by the modes, the difference between the standard and express mode
avg_pickup_delta = avg_standard_pickup - _____

<details>

<summary>Spoiler</summary>

```python

# calculate the delta for pickup_time by the modes
avg_pickup_delta = avg_standard_pickup - avg_express_pickup

```
</details>

#### Step 3 : Create new column taking into account the delta

In [70]:
# Subtract the value only for non-Express shipping method
pickup_time_kpi["expressed_no_weekend"] = np.where(pickup_time_kpi['_____'] != 'Express', #select orders that the ship_mode is not express
                                          pickup_time_kpi["pickup_time_delta_no_weekend"] - ______, # subtract the delta for those orders
                                          pickup_time_kpi["pickup_time_delta_no_weekend"]) # keep the express orders values as they

<details>

<summary>Spoiler</summary>

```python

# Subtract the value only for non-Express shipping method
pickup_time_kpi["expressed_no_weekend"] = np.where(pickup_time_kpi['ship_mode'] != 'Express',
                                          pickup_time_kpi["pickup_time_delta_no_weekend"] - avg_pickup_delta,
                                          pickup_time_kpi["pickup_time_delta_no_weekend"])

```
</details>

In [71]:
# display the following columns for the pickup_time_df ["expressed_no_weekend","pickup_time_delta_no_weekend","ship_mode"]
pickup_time_kpi[["expressed_no_weekend","pickup_time_delta_no_weekend","ship_mode"]]

expressed_no_weekend  pickup_time_delta_no_weekend            ship_mode
0                1.000000                             1              Express
1                0.000000                             0              Express
2                0.000000                             0              Express
3               -0.299138                             1  Standard Processing
4                1.000000                             1              Express
..                    ...                           ...                  ...
199              1.000000                             1              Express
200             -0.299138                             1  Standard Processing
201             -0.299138                             1  Standard Processing
202             -0.299138                             1  Standard Processing
203              0.700862                             2  Standard Processing

[204 rows x 3 columns]

We once again have instances where the newly calculated column has negatives, which would not make any sense. So we will set the value to 1.

In [72]:
# we take the maximum value for each row or replace it with 1 if less than 1 
pickup_time_kpi["expressed_no_weekend"] = np.maximum(pickup_time_kpi["expressed_no_weekend"], 1)

In [73]:
# display the following columns for the pickup_time_df ["expressed_no_weekend","pickup_time_delta_no_weekend","ship_mode"]
pickup_time_kpi[["___________","pickup_time_delta_no_weekend","ship_mode"]]

expressed_no_weekend  pickup_time_delta_no_weekend            ship_mode
0                     1.0                             1              Express
1                     1.0                             0              Express
2                     1.0                             0              Express
3                     1.0                             1  Standard Processing
4                     1.0                             1              Express
..                    ...                           ...                  ...
199                   1.0                             1              Express
200                   1.0                             1  Standard Processing
201                   1.0                             1  Standard Processing
202                   1.0                             1  Standard Processing
203                   1.0                             2  Standard Processing

[204 rows x 3 columns]

<details>

<summary>Spoiler</summary>

```python

# display the following columns for the pickup_time_df ["expressed_no_weekend","pickup_time_delta_no_weekend","ship_mode"]
pickup_time_kpi[["expressed_no_weekend","pickup_time_delta_no_weekend","ship_mode"]]

```
</details>

#### Step 4: The new KPI based on only Express ship_mode

In [74]:
# The new average order_cycle if only express using the column "expressed_no_weekend"
new_pickup_time_kpi = pickup_time_kpi["_____"].____

1.0461291476480634

<details>

<summary>Spoiler</summary>

```python

# The new average order_cycle if only express using the column "expressed_no_weekend"
pickup_time_kpi["expressed_no_weekend"].mean()

```
</details>

In [ ]:
new_pickup_time_kpi 

|  **KPIs**                 | **Warehouse Assumption**|  **Without Weekend**|**Prescription Express** |
|  -----------              |      -----------        |              ----      |        ----             |
|  Order Cycle Time         |          2 days         |         2.80 days  |        1.72 days       |
|       Pickup Time         |          1 day          |        1.35 days |      1.05 days       |


# Delivery time KPI and ship mode


The final kpi to explore with just the express shipping mode

#### Step 1: Splitting by the Ship Mode & Getting Averages

In [75]:
# express_shipping is a subset of all orders in the shipping_kpi dataframe that contains express orders only
express_shipping = shipping_kpi[shipping_kpi["______"] == "____"]

<details>

<summary>Spoiler</summary>

```python

# express_shipping is a subset of all orders in the shipping_kpi dataframe that contains express orders only
express_shipping = shipping_kpi[shipping_kpi["ship_mode"] == "Express"]

```
</details>

In [76]:
# standard_shipping is a subset of all orders in the shipping_kpi dataframe that contains the rest of the orders
standard_shipping = ______[______["_____"] != "Express"]

<details>

<summary>Spoiler</summary>

```python

# standard_shipping is a subset of all orders in the shipping_kpi dataframe that contains the rest of the orders
standard_shipping = shipping_kpi[shipping_kpi["ship_mode"] != "Express"]

```
</details>

#### The averages for each ship mode

In [79]:
# averge delivery times for standard
avg_standard_shipping = standard_shipping.package_delivered_int_no_weekend.mean()

2.941860465116279

In [80]:
# average delivery times for express 
avg_express_shipping = express_shipping.package_delivered_int_no_weekend.mean()

2.8533333333333335

#### Step 2 : The delta between each ship mode average

In [82]:
avg_shipping_delta = _____ - avg_express_shipping

<details>

<summary>Spoiler</summary>

```python

avg_shipping_delta = avg_standard_shipping - avg_express_shipping
               
```
</details>

#### Step 3 : Create new column taking into account the delta


In [83]:
# Subtract the value only for non-Express shipping method
shipping_kpi["expressed_no_weekend"] = np.where(_______['ship_mode'] != 'Express', #select orders that the ship_mode is not express
                                       _______["package_delivered_int_no_weekend"] - ______, # subtract the delta for those orders
                                       _______["package_delivered_int_no_weekend"]) # keep the express values as they are


expressed_no_weekend  package_delivered_int_no_weekend  \
0                4.000000                                 4   
1                2.700862                                 4   
2                3.700862                                 5   
3                1.700862                                 3   
4                2.700862                                 4   
..                    ...                               ...   
328              3.700862                                 5   
329              2.700862                                 4   
330              2.000000                                 2   
331              1.000000                                 2   
332              1.000000                                 2   

               ship_mode  
0                Express  
1    Standard Processing  
2    Standard Processing  
3    Standard Processing  
4    Standard Processing  
..                   ...  
328  Standard Processing  
329  Standard Processing  
330              Express  
331  Standard Processing  
332  Standard Processing  

[333 rows x 3 columns]

<details>

<summary>Spoiler</summary>

```python

# Subtract the value only for non-Express shipping method
shipping_kpi["expressed_no_weekend"] = np.where(shipping_kpi['ship_mode'] != 'Express', #select orders that the ship_mode is not express
                                       shipping_kpi["package_delivered_int_no_weekend"] - avg_shipping_delta, # subtract the delta for those orders
                                       shipping_kpi["package_delivered_int_no_weekend"]) # keep the express values as they are
               
```
</details>

In [ ]:
# we take the maximum value for each row in the new column "expressed_no_weekend" or replace it with 1 if less than 1 
shipping_kpi["expressed_no_weekend"] = np.maximum(shipping_kpi["______"], 1)


<details>

<summary>Answer</summary>

```python

# we take the maximum value for each row in the new column "expressed_no_weekend" or replace it with 1 if less than 1 
shipping_kpi["expressed_no_weekend"] = np.maximum(shipping_kpi["expressed_no_weekend"], 1)
               
```
</details>

In [ ]:
# displaying the columns to get an overview
shipping_kpi[["expressed_no_weekend","package_delivered_int_no_weekend","ship_mode"]]

#### Step 4: The new KPI based on only Express ship_mode

In [84]:
new_shipping_kpi = shipping_kpi["expressed_no_weekend"].mean()

2.0251579831737274

In [ ]:
new_shipping_kpi

|  **KPIs**                 | **Warehouse Assumption**|  **Without Weekend**|**Prescription Express** |
|  -----------              |      -----------        |              ----      |        ----             |
|  Order Cycle Time         |          2 days         |         2.80 days  |        1.72 days       |
|       Pickup Time         |          1 day          |        1.35 days |      1.05 days       |
| Delivery Time             |          3 days         |         2.92 days   |       2.03 days       |


# Bringing it all together

In [85]:
# Order_cycle_time + pickup_time + delivery_time
new_order_lead_time_kpi = new_order_cycle_kpi +  new_pickup_time_kpi + new_shipping_kpi 

In [86]:
new_order_lead_time_kpi

4.8

# Final outlook on the KPIs 

|  **KPIs**                 | **Warehouse Assumption**| **Data Driven Reality**| **Without Weekend**|**Prescription Daily Trucks**|**Prescription Express** |
|  -----------              |      -----------        |        -----------     |       ----         |    ----                     |        ----             |
|  Order Cycle Time         |          2 days         |       4.18 days        |         2.80 days  | 2.80 days                            |         1.72 days       |
|       Pickup Time         |          1 day          |       1.57 days        |          1.35 days |0.74 days|         1.05 days       |
| Delivery Time             |          3 days         |       4.6 days         |        2.92 days   | 2.92 days                            |         2.03 days       |
|     Order Lead Time       |  2 + 3(delivery) =5 days|       10.87 days       |          7.39 days |        6.46 days                     |         4.8 days        |


You have now managed to:

- Explore the data & Prepped it.

- Perform Descriptive Analytics : the WHAT

- Perform Diagnostic Analytics : the WHY

- Realized it is an interative process for all the above

- And now Precriptive Analytics.

**What's left?** 

Well bringing it all together, cherry pick the graphs that you think will best showcase and emphasize your findings and help in telling your story 